power_grid_model_c/power_grid_model_cpp/include/power_grid_model_cpp/serialization.hpp

```diff
@@ -0,0 +1,77 @@
+// SPDX-FileCopyrightText: Contributors to the Power Grid Model project <powergridmodel@lfenergy.org>
+//
+// SPDX-License-Identifier: MPL-2.0
+
+#pragma once
+#ifndef POWER_GRID_MODEL_CPP_SERIALIZATION_HPP
+#define POWER_GRID_MODEL_CPP_SERIALIZATION_HPP
+
+#include "basics.hpp"
+#include "dataset.hpp"
+#include "handle.hpp"
+
+#include "power_grid_model_c/serialization.h"
+
+namespace power_grid_model_cpp {
+class Deserializer {
+  public:
+    Deserializer(std::vector<char> const& data, Idx serialization_format)
+        : deserializer_{handle_.call_with(PGM_create_deserializer_from_binary_buffer, data.data(),
+                                          static_cast<Idx>(data.size()), serialization_format)} {}
+    Deserializer(std::string const& data_string, Idx serialization_format)
+        : deserializer_{handle_.call_with(PGM_create_deserializer_from_null_terminated_string, data_string.c_str(),
+                                          serialization_format)} {}
+
+    RawDeserializer* get() const { return deserializer_.get(); }
+
+    static DatasetWritable get_dataset(Deserializer const& deserializer) {
+        return DatasetWritable{deserializer.handle_.call_with(PGM_deserializer_get_dataset, deserializer.get())};
+    }
+    DatasetWritable get_dataset() const { return get_dataset(*this); }
+
+    static void parse_to_buffer(Deserializer& deserializer) {
+        deserializer.handle_.call_with(PGM_deserializer_parse_to_buffer, deserializer.get());
+    }
+    void parse_to_buffer() { parse_to_buffer(*this); }
+
+  private:
+    Handle handle_{};
+    detail::UniquePtr<RawDeserializer, PGM_destroy_deserializer> deserializer_;
+};
+
+class Serializer {
+  public:
+    Serializer(DatasetConst const& dataset, Idx serialization_format)
+        : serializer_{handle_.call_with(PGM_create_serializer, dataset.get(), serialization_format)} {}
+
+    RawSerializer* get() const { return serializer_.get(); }
+
+    static void get_to_binary_buffer(Serializer& serializer, Idx use_compact_list, std::vector<std::byte>& data) {
+        char const* temp_data{};
+        Idx buffer_size{};
+        serializer.handle_.call_with(PGM_serializer_get_to_binary_buffer, serializer.get(), use_compact_list,
+                                     &temp_data, &buffer_size);
+        if (temp_data != nullptr) {
+            data.resize(buffer_size);
+            std::memcpy(data.data(), temp_data, buffer_size);
+        }
+    }
+    void get_to_binary_buffer(Idx use_compact_list, std::vector<std::byte>& data) {
+        get_to_binary_buffer(*this, use_compact_list, data);
+    }
```

Comment:
i think it makes more sense to just output the `std::span<std::byte const>` directly. The user doesn't own the data themselves (in particular, the serializer returns a pointer to its data), but it is still the users' choice whether to use it directly or whether to copy it. In addition, the user may not want to provide you with an `std::vector` at all, but with some custom data type

In addition, `get_to` is a little bit cryptic here, because you don't actually specify where the serializer will write its data to in the C++ interface.

What you can do is maybe the following

```suggestion
    static std::span<std::byte const> to_binary_buffer(Serializer& serializer, Idx use_compact_list) {
        char const* temp_data{};
        Idx buffer_size{};
        serializer.handle_.call_with(PGM_serializer_get_to_binary_buffer, serializer.get(), use_compact_list,
                                     &temp_data, &buffer_size);
        return {temp_data, buffer_size};
    }
    std::span<std::byte const> to_binary_buffer(Idx use_compact_list) {
        return to_binary_buffer(*this, use_compact_list);
    }
    template <typename ByteContainer> static void get_to_binary_buffer(Serializer& serializer, Idx use_compact_list, ByteContainer& data) {
        std::span<std::byte const> buffer = to_binary_buffer(serializer, use_compact_list);
        if (buffer == nullptr) {
            data = {};
        } else {
            data.resize(buffer_size);
            std::memcpy(data.data(), temp_data, buffer_size);
        }
    }
    void get_to_binary_buffer(Idx use_compact_list, std::vector<std::byte>& data) {
        get_to_binary_buffer(*this, use_compact_list, data);
    }
```

Reply:
A slightly different implementation was used. Refer to commit [9c7b281](https://github.com/PowerGridModel/power-grid-model/pull/688/commits/9c7b281845a53f012f6d86553032a12075a46613) to see it.